server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/ReplacingServiceImpl.java

```diff
@@ -209,12 +213,90 @@ private String replace(@NonNull String text, @NonNull Pair<String, String> patte
+      result = String.join(System.lineSeparator(), fixedStrings);
     } catch (IndexOutOfBoundsException e) {
-      LOG.error(format(ERROR_REPLACING, text, pattern.toString()), e);
+      LOG.error(format(ERROR_REPLACING, text, pattern), e);
     }
     return result;
   }
 
+  private List<String> validateAndFix(String result, String referenceText) {
+    List<String> referenceStringList =
+        new ArrayList<>(Arrays.asList(referenceText.split(System.lineSeparator())));
+    List<String> resultStringList =
+        new ArrayList<>(Arrays.asList(result.split(System.lineSeparator())));
+    int resultIndex = 0;
+    for (int refIndex = 0; refIndex < referenceStringList.size(); refIndex++) {
+      int lengthDiff =
+          resultStringList.get(resultIndex).length() - referenceStringList.get(refIndex).length();
+
+      if (lengthDiff == 0) {
+        resultIndex++;
+        continue;
+      }
+      if (referenceStringList.get(refIndex).length() > END_INDEX_AREA_B
+          || resultStringList.get(resultIndex).length() > END_INDEX_AREA_B) {
+        String sequence = referenceStringList.get(refIndex).substring(END_INDEX_AREA_B);
+        if (lengthDiff < 0 && StringUtils.isNumeric(sequence)) {
+          resultStringList.set(
+              resultIndex, adjustSequenceNoPresentInAreaB(resultStringList, resultIndex, sequence));
+        } else {
+          int prevSize = resultStringList.size();
+          adjustExtraLength(resultStringList, refIndex, resultIndex, lengthDiff, sequence);
+          resultIndex = resultIndex + resultStringList.size() - prevSize;
+        }
+      }
+      resultIndex++;
+    }
+    return resultStringList;
+  }
+
+  private String adjustSequenceNoPresentInAreaB(
+      List<String> resultStringList, int j, String sequence) {
+    return StringUtils.join(
+        StringUtils.rightPad(
+            StringUtils.substring(
+                resultStringList.get(j), 0, resultStringList.get(j).length() - sequence.length()),
+            END_INDEX_AREA_B),
+        sequence);
+  }
+
+  private void adjustExtraLength(
+      List<String> resultStringList,
+      int refIndex,
+      int resultIndex,
+      int lengthDiff,
+      String sequence) {
+    String addedString =
+        resultStringList
+            .get(resultIndex)
+            .substring(END_INDEX_AREA_B, END_INDEX_AREA_B + lengthDiff - 1);
+    resultStringList.set(
+        refIndex,
+        StringUtils.join(
+            StringUtils.substring(resultStringList.get(resultIndex), 0, END_INDEX_AREA_B),
+            sequence));
+    if (!StringUtils.isBlank(addedString)) {
+      List<String> adjustedExtraCharInContinuationLine =
+          splitStringByLength(
+              addedString, END_INDEX_AREA_B - ProcessingConstants.CONTINUATION_PREFIX.length());
+      resultStringList.addAll(resultIndex + 1, adjustedExtraCharInContinuationLine);
+    }
+  }
+
+  private List<String> splitStringByLength(String actualStr, int fixedLength) {
```

Comment:
Here you change a number of strings and I'm afraid that this will change our mapping. I tried to reproduce the case when we need it and found this bug.
COBOL program:
```
0      IDENTIFICATION DIVISION.
1      PROGRAM-ID. TESTREPL.
2      DATA DIVISION.
3      WORKING-STORAGE SECTION.
       01  LOGA.  
       COPY REPL5  REPLACING ==LDAY== BY ==DMAN12345789sadfasfdsafasdfsa
      -    asdfsafsfabfgbtdndfrg==.

       PROCEDURE DIVISION.
           MOVE 0 TO DMAN12345789sadfasfdsafasdfsaasdfsafsfabfgbtdndfrg.
```
COPYBOOK REPL5.
```
      ***************************************************************** 09700000
         02  LOGHDR.                                                    18000000
            03  LDAY           PIC S9(7) COMP-3.    
```
I see that after the preprocessor changed line with `03  LDAY` is very long and not split. This also causes some errors to the user. Did you try to solve issues like this?

Reply:
Well, I tried something like this. I actually hoped that it won't give any mapping issues as the replacement with continued lines happens before we start the actual pre-processing. And the long replacement text is should be because the pre-processor concatenates.  But I guess, I missed some scenarios. I will look into it. Thanks :)
![image](https://user-images.githubusercontent.com/69206564/123413552-8b2f9f80-d5d0-11eb-8d4d-d116fe3dfff1.png)

Reply:
Hmm. That's interesting because you are showing the exact case I did, but you don't have any errors.

Reply:
It was my mistake. I have added a use-case test for this scenario.